src/QueryPipeline/QueryPipelineBuilder.cpp

```diff
@@ -618,6 +619,31 @@ void QueryPipelineBuilder::setProgressCallback(ProgressCallback callback)
     progress_callback = callback;
 }
 
+void QueryPipelineBuilder::connectDependencies()
+{
+    std::vector<RemoteSource *> input_dependencies;
+    std::vector<ReadFromMergeTreeDependencyTransform *> output_dependencies;
+
+    for (auto & processor : *pipe.getProcessorsPtr())
+    {
+        if (auto * remote_dependency = dynamic_cast<RemoteSource *>(processor.get()); remote_dependency)
+            input_dependencies.emplace_back(remote_dependency);
+        if (auto * merge_tree_dependency = dynamic_cast<ReadFromMergeTreeDependencyTransform *>(processor.get()); merge_tree_dependency)
```

Comment:
maybe make `RemoteSource` and `ReadFromMergeTreeDependencyTransform` final classes, and use typeid_cast here ...